File: Dynamic_Programming/01KnapSack/count_subset_sum.java
class Solution{

    //brute force sol

    public static int count_subset_sum(int [] nums, int sum) {

        // for DP

        Integer [][] dp = new Integer[nums.length][sum];

        return countSubsetDP(dp, nums, sum, 0);

    }


    private static int countSubsetRecur(int [] nums, int sum, int currentIndex) {
        if(sum == 0) {
            return 1;
        }

        if(nums.length == 0 || currentIndex >= nums.length) {
            return 0;
        }

        int sum1 = 0;

        if(nums[currentIndex] <= sum) {
            sum1 = countSubsetRecur(nums, sum - nums[currentIndex], currentIndex + 1);
        }

        //recursively call after excluding the number at the currentIndex
        int sum2 = countSubsetRecur(nums, sum, currentIndex + 1);

        return sum1 + sum2;
    }



    public static int countSubsetDP(Integer [][] dp, int [] nums, int sum, int currentIndex) {


        if (sum == 0) {
            return 1;
        }

        if (nums.length == 0 || currentIndex >= nums.length) {
            return 0;
        }
        if(dp[currentIndex][sum] == null) { 
            int sum1 = 0;

            if(nums[currentIndex] <= sum) {
                sum1 = dp[currentIndex][sum] = countSubsetDP(dp, nums, sum - nums[currentIndex], currentIndex + 1);
            }

            int sum2 = countSubsetDP(dp, nums, sum, currentIndex + 1);

            dp[currentIndex][sum] = sum1 + sum2;
        }

        //return dp[currentIndex][sum];

        return dp[currentIndex][sum];

        

    }
}

File: EDUCATIVE/Dynamic_Programming/01KnapSack/count_subset_sum.py
def count_subsets(num, sum):
    return count_subset_sum_recursive(num, sum, 0)

def count_subset_sum_recursive(num, sum, currentIndex):
    if sum == 0:
        return 1

    n = len(num)
    if n == 0 or currentIndex >= n:
        return 0

    sum1 = 0
    #include the item into the list
    if num[currentIndex] <= sum:
        sum1 = count_subset_sum_recursive(num, sum - num[currentIndex], currentIndex + 1)

    sum2 = count_subset_sum_recursive(num, sum, currentIndex + 1)

    return sum1 + sum2

def main():
  print("Total number of subsets " + str(count_subsets([1, 1, 2, 3], 4)))
  print("Total number of subsets: " + str(count_subsets([1, 2, 7, 1, 5], 9)))


main()